rust/extns/numaflow-kafka/src/lib.rs

```diff
@@ -0,0 +1,680 @@
+use std::{collections::HashMap, time::Duration};
+
+use bytes::Bytes;
+use rdkafka::Offset;
+use rdkafka::client::ClientContext;
+use rdkafka::config::{ClientConfig, RDKafkaLogLevel};
+use rdkafka::consumer::stream_consumer::StreamConsumer;
+use rdkafka::consumer::{BaseConsumer, CommitMode, Consumer, ConsumerContext, Rebalance};
+use rdkafka::error::KafkaResult;
+use rdkafka::message::{Headers, Message};
+use rdkafka::topic_partition_list::TopicPartitionList;
+use tokio::sync::{mpsc, oneshot};
+use tracing::{error, info, warn};
+
+pub type Result<T> = core::result::Result<T, Error>;
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("Connecting to Kafka {server} - {error}")]
+    Connection { server: String, error: String },
+
+    #[error("Kafka - {0}")]
+    Kafka(String),
+
+    #[error("{0}")]
+    Other(String),
+}
+
+/// Represents the authentication method used to connect to Kafka.
+#[derive(Debug, Clone, PartialEq)]
+pub enum KafkaAuth {
+    Sasl {
+        mechanism: String,
+        username: String,
+        password: String,
+    },
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct TlsConfig {
+    pub insecure_skip_verify: bool,
+    pub ca_cert: Option<String>,
+    pub client_auth: Option<TlsClientAuthCerts>,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct TlsClientAuthCerts {
+    pub client_cert: String,
+    pub client_cert_private_key: String,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct KafkaSourceConfig {
+    pub brokers: Vec<String>,
+    pub topic: String,
+    pub consumer_group: String,
+    pub auth: Option<KafkaAuth>,
+    pub tls: Option<TlsConfig>,
+}
+
+/// Message represents a message received from Kafka which can be converted to Numaflow Message.
+#[derive(Debug)]
+pub struct KafkaMessage {
+    /// The user payload.
+    pub value: Bytes,
+    /// The partition number.
+    pub partition: i32,
+    /// The offset of the message.
+    pub offset: i64,
+    pub headers: HashMap<String, String>,
+}
+
+// A context can be used to change the behavior of consumers by adding callbacks
+// that will be executed by librdkafka.
+struct KafkaContext;
+
+impl ClientContext for KafkaContext {}
+
+impl ConsumerContext for KafkaContext {
+    fn pre_rebalance(&self, _: &BaseConsumer<Self>, rebalance: &Rebalance<'_>) {
+        info!("Pre rebalance {:?}", rebalance);
+    }
+
+    fn post_rebalance(&self, _: &BaseConsumer<Self>, rebalance: &Rebalance<'_>) {
+        info!("Post rebalance {:?}", rebalance);
+    }
+
+    fn commit_callback(&self, result: KafkaResult<()>, _offsets: &TopicPartitionList) {
+        info!("Committing offsets: {:?}", result);
+    }
+}
+
+type LoggingConsumer = StreamConsumer<KafkaContext>;
+
+enum KafkaActorMessage {
+    Read {
+        respond_to: oneshot::Sender<Result<Vec<KafkaMessage>>>,
+    },
+    Ack {
+        offsets: Vec<(i32, i64)>,
```

Comment:
```suggestion
        offsets: Vec<(u32, u64)>,
```

Reply:
I'm using the same data types as rdkafka library.